plugins/amqp_trx_plugin/amqp_trx_plugin.cpp

```diff
@@ -0,0 +1,169 @@
+         if( trace_plug ) {
+            std::string err = "Dropped trx, too many trx in progress " + std::to_string( trx_in_progress ) + " bytes";
+            trace_plug->publish_error( trx->id().str(), chain::tx_resource_exhaustion::code_enum::code_value, std::move(err) );
+         }
+         return;
+      }
+
+      trx_in_progress_size += trx->get_estimated_size();
+      app().post( priority::medium_low, [my=shared_from_this(), trx{std::move(trx)}]() {
+         my->chain_plug->accept_transaction( trx,
+            [my, trx](const fc::static_variant<fc::exception_ptr, chain::transaction_trace_ptr>& result) mutable {
+               // publish to trace plugin as execptions are not reported via controller signal applied_transaction
+               if( result.contains<chain::exception_ptr>() && my->trace_plug ) {
+                  auto& eptr = result.get<chain::exception_ptr>();
+                  my->trace_plug->publish_error( trx->id().str(), eptr->code(), eptr->to_string() );
+               }
+               my->trx_in_progress_size -= trx->get_estimated_size();
+            } );
+         } );
+   }
+};
+
+amqp_trx_plugin::amqp_trx_plugin()
+: my(std::make_shared<amqp_trx_plugin_impl>()) {}
+
+amqp_trx_plugin::~amqp_trx_plugin() {}
+
+void amqp_trx_plugin::set_program_options(options_description& cli, options_description& cfg) {
+   auto op = cfg.add_options();
+   op("amqp-trx-address", bpo::value<std::string>(),
+      "AMQP address: Format: amqp://USER:PASSWORD@ADDRESS:PORT\n"
+      "Will consume from 'trx' queue.");
+}
+
+void amqp_trx_plugin::plugin_initialize(const variables_map& options) {
+   try {
+      my->chain_plug = app().find_plugin<chain_plugin>();
+      EOS_ASSERT( my->chain_plug, chain::missing_chain_plugin_exception, "chain_plugin required" );
+
+      my->trace_plug = app().find_plugin<amqp_trace_plugin>(); // optional
+
+      EOS_ASSERT( options.count("amqp-trx-address"), chain::plugin_config_exception, "amqp-trx-address required" );
+      my->amqp_trx_address = options.at("amqp-trx-address").as<std::string>();
+   }
+   FC_LOG_AND_RETHROW()
+}
+
+void amqp_trx_plugin::plugin_startup() {
+   handle_sighup();
+   try {
+
+      ilog( "Starting amqp_trx_plugin" );
+      my->thread_pool.emplace( "amqp_t", 1 );
+
+      my->amqp_trx.emplace( my->thread_pool->get_executor(), my->amqp_trx_address, "trx", [](const std::string& err) {
+         elog( "amqp error: ${e}", ("e", err) );
+         app().quit();
+      });
+
+      auto& consumer = my->amqp_trx->consume();
+      consumer.onSuccess( []( const std::string& consumer_tag ) {
+         dlog( "consume started: ${tag}", ("tag", consumer_tag) );
+      } );
+      consumer.onError( []( const char* message ) {
+         wlog( "consume failed: ${e}", ("e", message) );
+      } );
+      consumer.onReceived( [this](const AMQP::Message& message, uint64_t delivery_tag, bool redelivered) {
```

Comment:
`this` is somewhat scary as the callback can survive the destruction of `amqp_trx_shutdown` (It can be preserved by the `post` from `handle_message`).  I _think_ it's still safe, because the callback will only be executed by the thread pool, which is joined in `plugin_shutdown`.

Reply:
Agree. I was counting on the thread_pool shutdown protection.